File: .github/repro-538/repro.rb
require 'socket'
require 'timeout'
h = ARGV[0]
fams = Addrinfo.getaddrinfo(h, 443, nil, :STREAM).map { |a| a.ipv6? ? 'v6' : 'v4' }.uniq
res = Hash.new(0)
5.times do
  pid = fork { $stderr.reopen('/dev/null'); Addrinfo.getaddrinfo(h, 443, nil, :STREAM); exit!(0) }
  begin
    _, s = Timeout.timeout(15) { Process.waitpid2(pid) }
    res[s.signaled? ? "CRASH" : "ok"] += 1
  rescue Timeout::Error
    # ruby >= 3.3 resolves on a helper thread, so the child can hang instead of dying
    Process.kill(:KILL, pid)
    Process.waitpid(pid)
    res["HANG"] += 1
  end
end
puts "#{RUBY_VERSION} #{h} #{fams} => #{res}"

File: .github/repro-538/repro.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <netdb.h>
#include <sys/wait.h>
static int look(const char *h){struct addrinfo *r=0;int e=getaddrinfo(h,"443",0,&r);if(r)freeaddrinfo(r);return e;}
int main(int c,char**v){
  const char *h = c>1?v[1]:"github.com";
  look(h);
  int sig=0,ok=0;
  for(int i=0;i<10;i++){
    pid_t p=fork();
    if(!p){ alarm(10); look(h); _exit(0);}
    int st; waitpid(p,&st,0);
    if(WIFSIGNALED(st)) {sig++; fprintf(stderr,"child sig %d\n",WTERMSIG(st));} else ok++;
  }
  printf("ok=%d signaled=%d\n",ok,sig); return 0;}
